perception/sub8_perception/nodes/follow_orange_pipes.py

```diff
@@ -4,315 +4,228 @@
-        '''
-        xy_position, q, height = self.get_tf(timestamp, get_rotation=True)
-
-        dir_vector = unit_vector(np.array([self.cam.cx(), self.cam.cy()]) - m_position)
-        cam_rotation = tf.transformations.euler_from_quaternion(q)[2] + np.pi / 2
-        print "MARKER - dir_vector:", dir_vector
-        print "MARKER - cam_rotation:", cam_rotation
-        dir_vector = np.dot(dir_vector, make_2D_rotation(cam_rotation))
-
-        # Calculate distance from middle of frame to marker in meters.
-        magnitude = self.calculate_visual_radius(height, second_point=m_position)
-        abs_position = np.append(xy_position + dir_vector[::-1] * magnitude, -SEARCH_DEPTH)
-
-        return abs_position
-
-    def calculate_visual_radius(self, height, second_point=None):
-        '''
-        Draws rays to find the radius of the FOV of the camera in meters.
-        It also can work to find the distance between two planar points some distance from the camera.
-        '''
-
-        mid_ray = np.array([0, 0, 1])
-
-        if second_point is None:
-            if self.cam.cy() < self.cam.cx():
-                second_point = np.array([self.cam.cx(), 0])
-            else:
-                second_point = np.array([0, self.cam.cy()])
-
-        edge_ray = unit_vector(self.cam.projectPixelTo3dRay(second_point))
-
-        # Calculate angle between vectors and use that to find r
-        theta = np.arccos(np.dot(mid_ray, edge_ray))
-        return np.tan(theta) * height
-
-    def calculate_marker_area(self, height):
-        '''
-        What we really don't want is to find markers that are on the edge since the direction and center of
-            the marker are off.
-        '''
-        MARKER_LENGTH = 1.22  # m
-        MARKER_WIDTH = .1524  # m
-
-        # Get m/px on the ground floor.
-        m = self.calculate_visual_radius(height)
-        if self.cam.cy() < self.cam.cx():
-            px = self.cam.cy()
-        else:
-            px = self.cam.cx()
-
-        m_px = m / px
-        marker_area_m = MARKER_WIDTH * MARKER_LENGTH
-        marker_area_px = marker_area_m / (m_px ** 2)
-
-        return marker_area_px
-
-def unit_vector(vect):
-    return vect / np.linalg.norm(vect)
-
-
-def make_2D_rotation(angle):
-    c, s = np.cos(angle), np.sin(angle)
-    return np.array([[c, -s],
-                     [s, c]], dtype=np.float32)
-
-
-def main(args):
-    redniFrekraM = MarkerFinder()
-    try:
-        rospy.spin()
-    except KeyboardInterrupt:
-        print("Shutting down")
-
+            res.pose.x = self.last2d[0][0]
+            res.pose.y = self.last2d[0][1]
+            res.pose.theta = self.last2d[1]
+            res.found = True
+        return res
+
+    def get_3d_pose(self, p):
+        i_points = np.array((p[0][0], p[1][0], p[2][0], p[3][0]),dtype=np.float)
+        retval, rvec, tvec =  cv2.solvePnP(PathMarkerFinder.PATH_MARKER, i_points, self.cam.intrinsicMatrix(), np.zeros((5,1)))
```

Comment:
Yes. Every competition bag I found showing path markers completely in frame for more than a second are included in the unit tests and this node passes the 2D test. I don't know the actual 3D pose to test to.